// @sanity/vision/lib/Vision.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _VisionContainer = _interopRequireDefault(require("./containers/VisionContainer"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
// Passes the given Sanity client to use down
// through context to child components
class Vision extends _react.default.PureComponent {
  getChildContext() {
    return {
      client: this.props.client,
      schema: this.props.schema
    };
  }
  render() {
    return /*#__PURE__*/_react.default.createElement(_VisionContainer.default, this.props);
  }
}
Vision.propTypes = {
  client: _propTypes.default.shape({
    config: _propTypes.default.func
  }).isRequired,
  schema: _propTypes.default.object
};
Vision.childContextTypes = {
  client: _propTypes.default.shape({
    config: _propTypes.default.func
  }).isRequired,
  schema: _propTypes.default.object
};
var _default = exports.default = Vision;
