// @sanity/vision/lib/containers/VisionContainer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _DelayedSpinner = _interopRequireDefault(require("../components/DelayedSpinner"));
var _VisionGui = _interopRequireDefault(require("../components/VisionGui"));
var _LoadingContainer = _interopRequireDefault(require("./LoadingContainer"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
// Loads the most basic data from a Sanity project
class VisionContainer extends _LoadingContainer.default {
  getSubscriptions() {
    return {
      datasets: {
        uri: '/datasets'
      }
    };
  }
  render() {
    var datasets = this.state.datasets || [];
    if (this.state.error) {
      var defaultDataset = this.context.client.config().dataset;
      datasets[0] = {
        name: defaultDataset
      };
    } else if (!this.hasAllData()) {
      return /*#__PURE__*/_react.default.createElement(_DelayedSpinner.default, null);
    }
    return /*#__PURE__*/_react.default.createElement(_VisionGui.default, _extends({}, this.state, this.props, {
      datasets: datasets
    }));
  }
}
var _default = exports.default = VisionContainer;
