// @sanity/vision/lib/components/QueryErrorDetails.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _ui = require("@sanity/ui");
var _QueryErrorDialog = require("./QueryErrorDialog.styled");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function QueryErrorDetails(props) {
  var details = _objectSpread(_objectSpread({}, props.error.details), mapToLegacyDetails(props.error.details));
  if (!details.line) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_QueryErrorDialog.ErrorCode, {
    size: 1
  }, "".concat(details.line, "\n").concat(dashLine(details.column, details.columnEnd))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 4
  }, /*#__PURE__*/_react.default.createElement(_QueryErrorDialog.ErrorCode, {
    size: 1
  }, "Line:   ".concat(details.lineNumber, "\nColumn: ").concat(details.column))));
}
function mapToLegacyDetails(details) {
  if (!details || typeof details.query !== 'string' || typeof details.start !== 'number') {
    return {};
  }
  var query = details.query,
    start = details.start,
    end = details.end;
  var lineStart = query.slice(0, start).lastIndexOf('\n') + 1;
  var lineNumber = (query.slice(0, lineStart).match(/\n/g) || []).length;
  var line = query.slice(lineStart, query.indexOf('\n', lineStart));
  var column = start - lineStart;
  var columnEnd = typeof end === 'number' ? end - lineStart : undefined;
  return {
    line,
    lineNumber,
    column,
    columnEnd
  };
}
function dashLine(column, columnEnd) {
  var line = '-'.repeat(column);
  var hats = "^".repeat(columnEnd ? columnEnd - column : 1);
  return "".concat(line).concat(hats);
}
QueryErrorDetails.propTypes = {
  error: _propTypes.default.shape({
    details: _propTypes.default.shape({
      line: _propTypes.default.string,
      lineNumber: _propTypes.default.number,
      column: _propTypes.default.number,
      start: _propTypes.default.number,
      end: _propTypes.default.number,
      query: _propTypes.default.string
    })
  })
};
var _default = exports.default = QueryErrorDetails;
