// @sanity/validation/lib/ValidationError.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
// Follows the same pattern as Rule and RuleClass. @see Rule
var ValidationError = class ValidationError {
  constructor(message) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _defineProperty(this, "message", void 0);
    _defineProperty(this, "paths", void 0);
    _defineProperty(this, "children", void 0);
    _defineProperty(this, "operation", void 0);
    this.message = message;
    this.paths = options.paths || [];
    this.children = options.children;
    this.operation = options.operation;
  }
  cloneWithMessage(msg) {
    return new ValidationError(msg, {
      paths: this.paths,
      children: this.children,
      operation: this.operation
    });
  }
};
var _default = ValidationError;
exports.default = _default;
